packages/kbn-server-route-repository/src/test_types.ts

```diff
@@ -0,0 +1,227 @@
+        serviceName: t.string,
+      }),
+    }),
+    handler: async ({ params }) => {
+      return {
+        yesParamsForMe: true,
+      };
+    },
+  })
+  .add({
+    endpoint: 'endpoint_with_optional_params',
+    params: t.partial({
+      query: t.partial({
+        serviceName: t.string,
+      }),
+    }),
+    handler: async ({ params }) => {
+      return {
+        someParamsForMe: true,
+      };
+    },
+  });
+
+type TestRepository = typeof repository;
+
+// EndpointOf should return all valid endpoints of a repository
+
+const validEndpoints: Array<EndpointOf<TestRepository>> = [
+  'endpoint_with_params',
+  'endpoint_without_params',
+  'endpoint_with_optional_params',
+];
+
+// @ts-expect-error Type '"this_endpoint_does_not_exist"' is not assignable to type '"endpoint_without_params" | "endpoint_with_params" | "endpoint_with_optional_params"'
+const invalidEndpoints: Array<EndpointOf<TestRepository>> = ['this_endpoint_does_not_exist'];
+
+// ReturnOf should return the return type of a request handler.
+
+const noParamsValid: ReturnOf<TestRepository, 'endpoint_without_params'> = {
+  noParamsForMe: true,
+};
+
+const noParamsInvalid: ReturnOf<TestRepository, 'endpoint_without_params'> = {
+  // @ts-expect-error type '{ paramsForMe: boolean; }' is not assignable to type '{ noParamsForMe: boolean; }'.
+  paramsForMe: true,
+};
+
+// RouteRepositoryClient
+
+type TestClient = RouteRepositoryClient<TestRepository, { timeout: number }>;
+
+const client: TestClient = {} as any;
+
+// It should respect any additional create options.
+
+// @ts-expect-error Property 'timeout' is missing
+client({
+  endpoint: 'endpoint_without_params',
+});
+
+client({
+  endpoint: 'endpoint_without_params',
+  timeout: 1,
+});
+
+// It does not allow params for routes without a params codec
+client({
+  endpoint: 'endpoint_without_params',
+  // @ts-expect-error Object literal may only specify known properties, and 'params' does not exist in type
+  params: {},
+  timeout: 1,
+});
+
+// It requires params for routes with a params codec
+client({
+  endpoint: 'endpoint_with_params',
+  params: {
+    // @ts-expect-error property 'serviceName' is missing in type '{}'
+    path: {},
+  },
+  timeout: 1,
+});
+
+// Params are optional if the codec has no required keys
+client({
+  endpoint: 'endpoint_with_optional_params',
+  timeout: 1,
+});
+
+// If optional, an error will still occur if the params do not match
+client({
+  endpoint: 'endpoint_with_optional_params',
+  timeout: 1,
+  params: {
+    // @ts-expect-error Object literal may only specify known properties, and 'path' does not exist in type
+    path: '',
+  },
+});
+
+// The return type is correctly inferred
+client({
+  endpoint: 'endpoint_with_params',
+  params: {
+    path: {
+      serviceName: '',
+    },
+  },
+  timeout: 1,
+}).then((res) => {
+  // @ts-expect-error Property 'noParamsForMe' is missing in type
+  const invalidType: {
+    noParamsForMe: boolean;
+  } = res;
```

Comment:
If you have access to `expect` you can double this as both a runtime and compile. And I find the intention a little more clear this way

```suggestion
  // @ts-expect-error Property 'noParamsForMe' is missing in return type
  expect(res.noParamsForMe).toBe(undefined)
```

Reply:
expect might not be the appropriate tool, but let me see if I can write some other utility function

Reply:
what about something like 
![image](https://user-images.githubusercontent.com/352732/113860833-c7f4b300-97a6-11eb-8f44-5630fe9b90e9.png)

Reply:
An assert generic is pretty cool 👍